internal/mode/static/state/store.go

```diff
@@ -17,9 +20,59 @@ type Updater interface {
 
 // objectStore is a store of client.Object
 type objectStore interface {
-	get(nsname types.NamespacedName) client.Object
+	get(objType client.Object, nsname types.NamespacedName) client.Object
 	upsert(obj client.Object)
-	delete(nsname types.NamespacedName)
+	delete(objType client.Object, nsname types.NamespacedName)
+}
+
+// ngfPolicyObjectStore is a store of policies.Policy.
+// A single store should be used to store all types of policies.Policy.
+type ngfPolicyObjectStore struct {
+	policies       map[graph.PolicyKey]policies.Policy
+	extractGVKFunc extractGVKFunc
+}
+
+// newNGFPolicyObjectStore returns a new ngfPolicyObjectStore.
+func newNGFPolicyObjectStore(
+	policies map[graph.PolicyKey]policies.Policy,
+	gvkFunc extractGVKFunc,
+) *ngfPolicyObjectStore {
+	return &ngfPolicyObjectStore{
+		policies:       policies,
+		extractGVKFunc: gvkFunc,
+	}
+}
+
+func (p *ngfPolicyObjectStore) get(objType client.Object, nsname types.NamespacedName) client.Object {
+	key := graph.PolicyKey{
+		NsName: nsname,
+		GVK:    p.extractGVKFunc(objType),
+	}
+
+	return p.policies[key]
+}
+
+func (p *ngfPolicyObjectStore) upsert(obj client.Object) {
+	key := graph.PolicyKey{
+		NsName: client.ObjectKeyFromObject(obj),
+		GVK:    p.extractGVKFunc(obj),
+	}
+
+	pol, ok := obj.(policies.Policy)
+	if !ok {
+		panic(fmt.Sprintf("expected NGF Policy, got %T", obj))
+	}
+
+	p.policies[key] = pol
+}
+
+func (p *ngfPolicyObjectStore) delete(objType client.Object, nsname types.NamespacedName) {
+	key := graph.PolicyKey{
+		NsName: nsname,
+		GVK:    p.extractGVKFunc(objType),
```

Comment:
On a delete event, can the GVK be extracted from the object properly?

Reply:
Yep, that's why I used the function instead of trying to access directly though the object.